### PR TITLE
Problem: fix-unlucky-tx command runs too slow

### PR DESCRIPTION
**Pull request by jsvisa**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest master? 
- [x] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [x] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 1]
```diff
@@ -147,47 +151,92 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+					}
 				}
+				return nil
 			}
 
-			return nil
+			go func() {
+				err := findBlock()
+				if err != nil {
+					cancel()
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:232]:
```suggestion
                                          fmt.Fprintln(os.Stderr, err)
					cancel()
```
Log the error?

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 2]
```diff
@@ -147,47 +151,93 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+			ctCtx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			for i := 0; i < concurrency; i++ {
+				wg.Add(1)
+				go func(wg *sync.WaitGroup) {
+					defer wg.Done()
+					for {
+						select {
+						case <-ctCtx.Done():
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:170]: should we select the `Done()` together with the `blockChain`?
it seems doing dead loop after finish the works.

Comment (yihuang) [reply cmd/cronosd/cmd/fix-unlucky-tx.go:170]: I guess we need to call the cancel no mater if `findBlock` returns error, right?

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 3]
```diff
@@ -147,47 +151,93 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+							return
+						default:
+						}
+						select {
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:174]:
```suggestion
```

I think we should select the Done() and the blockChan together?

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 4]
```diff
@@ -147,47 +151,90 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+						case blockNum, ok := <-blockChan:
+							if !ok {
+								return
+							}
+
+							if err := processBlock(blockNum); err != nil {
+								cancel()
+								return
+							}
+						default:
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:181]:
```suggestion
```

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 5]
```diff
@@ -147,47 +151,90 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+					}
 				}
+				return nil
 			}
 
-			return nil
+			go func() {
+				err := findBlock()
+				if err != nil {
+                                        fmt.Fprintln(os.Stderr, err)
```
Comment (yihuang) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:229]:
```suggestion
                                       fmt.Fprintln(os.Stderr, err)
```

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 6]
```diff
@@ -147,47 +151,93 @@ func FixUnluckyTxCmd() *cobra.Command {
...
+					cancel()
+				}
```
[the remark this answers is not in the document]
Comment (jsvisa) [reply cmd/cronosd/cmd/fix-unlucky-tx.go]: context was canceled with defer

[the remark this answers is not in the document]
Comment (jsvisa) [reply cmd/cronosd/cmd/fix-unlucky-tx.go]: Sorry, the blockChan was not closed.

**Comment by jsvisa** [comment]
PTAL

**Review by yihuang — APPROVED** [review]
LGTM

**Comment by thomas-nguy** [comment]
@jsvisa could you sign your commits?

File: cmd/cronosd/cmd/fix-unlucky-tx.go [hunk 7]
```diff
@@ -196,6 +241,7 @@ func FixUnluckyTxCmd() *cobra.Command {
 	cmd.Flags().String(FlagBlocksFile, "", "Read block numbers from a file instead of iterating all the blocks")
 	cmd.Flags().Int(FlagStartBlock, 1, "The start of the block range to iterate, inclusive")
 	cmd.Flags().Int(FlagEndBlock, -1, "The end of the block range to iterate, inclusive")
+	cmd.Flags().Int(FlagConcurrency, runtime.NumCPU(), "Run how much workers in concurrency")
```
Comment (thomas-nguy) [inline cmd/cronosd/cmd/fix-unlucky-tx.go:244]:
```suggestion
	cmd.Flags().Int(FlagConcurrency, runtime.NumCPU(), "Define how much workers run in concurrency")
```

**Comment by jsvisa** [comment]
> @jsvisa could you sign your commits?

wait for a moment

**Comment by jsvisa** [comment]
> > @jsvisa could you sign your commits?
> 
> wait for a moment

Sorry, My GPG-Suite seems not working.

**Comment by yihuang** [comment]
> > > @jsvisa could you sign your commits?
> > 
> > 
> > wait for a moment
> 
> Sorry, My GPG-Suite seems not working.

if you click allow edits from the maintainer, I can push a squashed commit.

**Comment by jsvisa** [comment]
yeah, that's selected

image

**Comment by jsvisa** [comment]
The GPG sign seems working

**Pull request merged**